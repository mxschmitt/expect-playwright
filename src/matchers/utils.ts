import type { Page, ElementHandle, Frame } from "playwright-core"
import { PageWaitForSelectorOptions } from "../../global"

export type ExpectInputType = Page | Frame | ElementHandle

const isElementHandle = (value: ExpectInputType): value is ElementHandle => {
  return value.constructor.name === "ElementHandle"
}

export const getFrame = async (value: ExpectInputType) => {
  return isElementHandle(value) ? value.contentFrame() : value
}

const isObject = (value: unknown) =>
  typeof value === "object" && !(value instanceof RegExp)

export type InputArguments = [
  Page | ElementHandle,
  string?,
  (string | PageWaitForSelectorOptions)?,
  PageWaitForSelectorOptions?
]

export const getElementHandle = async (
  args: InputArguments,
  valueArgCount = 1
) => {
  // Pluck the options off the end first
  const options =
    args.length > 1 && isObject(args[args.length - 1])
      ? (args.pop() as PageWaitForSelectorOptions)
      : {}

  // Next, pluck the number of args required by the matcher (defaults to 1)
  const expectedValue = args.splice(-valueArgCount, valueArgCount) as string[]

  // Finally, we can find the element handle
  let elementHandle = (await getFrame(args[0])) ?? args[0]

  // If the user provided a page or iframe, we need to locate the provided
  // selector or the `body` element if none was provided.
  if (!isElementHandle(elementHandle)) {
    const selector = args[1] ?? "body"

    try {
      elementHandle = (await elementHandle.waitForSelector(selector, options))!
    } catch (err) {
      throw new Error(`Timeout exceed for element ${quote(selector)}`)
    }
  }

  return [elementHandle, expectedValue] as const
}

export const quote = (val: string | null) => (val === null ? "" : `'${val}'`)

export const getMessage = (
  { isNot, promise, utils, expand }: jest.MatcherContext,
  matcher: string,
  expected: unknown,
  received: unknown,
  expectedHint: string | undefined = undefined
) => {
  const message = isNot
    ? `Expected: not ${utils.printExpected(expected)}`
    : utils.printDiffOrStringify(
        expected,
        received,
        "Expected",
        "Received",
        expand
      )

  return (
    utils.matcherHint(matcher, undefined, expectedHint, { isNot, promise }) +
    "\n\n" +
    message
  )
}

export const compareText = (
  expectedValue: string | RegExp,
  actualValue: string | null
) => {
  return typeof expectedValue === "string"
    ? expectedValue === actualValue
    : expectedValue.test(actualValue ?? "")
}
