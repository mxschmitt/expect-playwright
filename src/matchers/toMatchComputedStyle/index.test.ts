import { assertSnapshot } from "../tests/utils"

const black = "rgb(0, 0, 0)"
const content = `
  <style>#foo { background-color: black }</style>
  <p id="foo">Hi</p>
`

describe("toMatchComputedStyle", () => {
  beforeEach(() => jestPlaywright.resetContext())

  describe("selector", () => {
    it("positive", async () => {
      await page.setContent(content)
      await expect(page).toMatchComputedStyle("#foo", "background-color", black)
      await expect(page).toMatchComputedStyle("#foo", "background-color", /rgb/)
    })

    it("positive in frame", async () => {
      await page.setContent('<iframe src="http://localhost:8080">')
      const args = ["#styles", "color", "rgb(255, 0, 0)"] as const

      const handle = page.$("iframe")
      await expect(handle).toMatchComputedStyle(...args)
      await expect(await handle).toMatchComputedStyle(...args)

      const iframe = (await handle)?.contentFrame()
      await expect(iframe).toMatchComputedStyle(...args)
      await expect(await iframe).toMatchComputedStyle(...args)
    })

    it("negative", async () => {
      await page.setContent(content)
      await assertSnapshot(() =>
        expect(page).toMatchComputedStyle("#foo", "color", "white")
      )
    })

    describe("with 'not' usage", () => {
      it("positive", async () => {
        await page.setContent(content)
        expect(page).not.toMatchComputedStyle("#foo", "color", "white")
        expect(page).not.toMatchComputedStyle("#foo", "color", /rgba/)
      })

      it("positive in frame", async () => {
        await page.setContent('<iframe src="http://localhost:8080">')
        const args = ["#styles", "color", "rgb(0, 0, 0)"] as const

        const handle = page.$("iframe")
        await expect(handle).not.toMatchComputedStyle(...args)
        await expect(await handle).not.toMatchComputedStyle(...args)

        const iframe = (await handle)?.contentFrame()
        await expect(iframe).not.toMatchComputedStyle(...args)
        await expect(await iframe).not.toMatchComputedStyle(...args)
      })

      it("negative", async () => {
        await page.setContent(content)
        await assertSnapshot(() =>
          expect(page).not.toMatchComputedStyle("#foo", "color", black)
        )
      })
    })
  })

  describe("element", () => {
    it("positive", async () => {
      await page.setContent(content)
      const element = page.$("#foo")
      await expect(element).toMatchComputedStyle("color", black)
      await expect(await element).toMatchComputedStyle("color", /rgb/)
    })

    it("negative", async () => {
      await page.setContent(content)
      const element = await page.$("#foo")
      await assertSnapshot(() =>
        expect(element).toMatchComputedStyle("background-color", "white")
      )
    })
  })

  describe("timeout", () => {
    it("should throw an error after the timeout exceeds", async () => {
      const start = new Date().getTime()
      await assertSnapshot(() =>
        expect(page).toMatchComputedStyle("#foo", "color", "white", {
          timeout: 1000,
        })
      )
      const duration = new Date().getTime() - start
      expect(duration).toBeGreaterThan(1000)
      expect(duration).toBeLessThan(1500)
    })
  })
})
