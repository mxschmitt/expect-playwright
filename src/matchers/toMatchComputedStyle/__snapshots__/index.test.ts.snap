// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toMatchComputedStyle element negative 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoMatchComputedStyle[2m([22m[32mexpected[39m[2m)[22m

Expected: [32m\\"white\\"[39m
Received: [31m\\"rgb(0, 0, 0)\\"[39m"
`;

exports[`toMatchComputedStyle selector negative 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoMatchComputedStyle[2m([22m[32mexpected[39m[2m)[22m

Expected: [32m\\"white\\"[39m
Received: [31m\\"rgb(0, 0, 0)\\"[39m"
`;

exports[`toMatchComputedStyle selector with 'not' usage negative 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mnot[2m.[22mtoMatchComputedStyle[2m([22m[32mexpected[39m[2m)[22m

Expected: not [32m\\"rgb(0, 0, 0)\\"[39m"
`;

exports[`toMatchComputedStyle timeout should throw an error after the timeout exceeds 1`] = `"Error: Timeout exceed for element '#foo'"`;
